codex/utils/asyncthreads.nim

```diff
@@ -0,0 +1,33 @@
+
+import pkg/taskpools
+import pkg/taskpools/flowvars
+import pkg/chronos
+import pkg/chronos/threadsync
+import pkg/questionable/results
+
+const
+  CompletionRetryDelay* = 10.millis
+  CompletionTimeout* = 1.seconds # Maximum await time for completition after receiving a signal
+
+proc awaitThreadResult*[T](signal: ThreadSignalPtr, handle: Flowvar[T]): Future[?!T] {.async.} =
```

Comment:
I know this exists in the erasure async backend right now, but imp this is way too convoluted. 

Maybe something like this:

```suggestion
proc awaitThreadResult*[T](signal: ThreadSignalPtr, handle: Flowvar[T]): Future[?!T] {.async.} =
  await wait(signal)
  success(handle.sync())
```

Reply:
>   `await wait(signal)`
>   `success(handle.sync())`

Unfortunately, this doesn't work properly because it can start running taskpool jobs on the current thread.

What happens is that the tasks have to call `fireSync` before they return and the taskpools worker can put the result into the `Flowvar` and mark it as finished. There's a chance here that the Chronos thread will receive the signal and proceed to call `sync` *before* the Flowvar is marked as completed. If, or probably when, this occurs it'll block the Chronos thread and cause lots of headaches.

That *shouldn't* happen normally, but it's a common race condition with this kinda worker pattern. With high load on a system one I'd expect to happen once in a while.

My preferred way to handle this case is to have our own queue to put the results in it and then call the signal. This way you avoid the race condition entirely.

Reply:
This only happens for nested tasks, but it might be a problem due to task stealing. 

I would avoid flowvar altogether and just rely on the chronos signal, otherwise we're mixing messaging primitives, even if we're doing `tryComplete/tryRecv`.

> My preferred way to handle this case is to have our own queue to put the results in it and then call the signal. This way you avoid the race condition entirely.

No need, taskpools is already a queue why add yet another just to pass around a (dubiously useful) message?

What I would do is:

- Remove the flowvar
- Add a simple bool flag to the thread args
  - If more complex error conditions need to be communicated, use an enum
- Use chronos signaling
- Log the error in the task itself

Reply:
> This only happens for nested tasks, but it might be a problem due to task stealing.

Yes looks like it'll try work stealing. @tbekas confirmed this previously in his testing for the erasure encoding threading.

> I would avoid flowvar altogether and just rely on the chronos signal, otherwise we're mixing messaging primitives, even if we're doing tryComplete/tryRecv.

The flowvar does provide convenience though. It's unforunate that just using it for returning data still requires checking `tryComplete`. 

> What I would do is:
> Remove the flowvar

If we don't use taskpools flowvar, then we need handle setting up the memory for where to place the proof results. It'd also mean we are doing threading in two different ways in the codebase. 

> Add a simple bool flag to the thread args
> If more complex error conditions need to be communicated, use an enum
> Log the error in the task itself

Using a result type seems simpler than setting up all of that. :)

> Use chronos signaling

This is using the Chronos signaling, same as the erasure coding threading. It's more about how we want to return the data.

Reply:
> No need, taskpools is already a queue why add yet another just to pass around a (dubiously useful) message?

To return the proof data (or erasure encoded data, etc) and a queue is just the canonical way to do that in worker pools. Taskpools uses the two queues pattern as well: the job queue, and the flowvar's. The flowvars are a `ChannelSPSCSingle` queue with fixed size.

My preferred approach would be to replace taskpool's flowvars (output queues) with a thread-async-aware queue that wraps ThreadSignalPtr and a standard locking queue so we don't need to re-invent the wheel each time. 

Then you just pass the results queue to the task and it avoids both the mixed signaling primitives and needing to decide how to allocate the results memory each time.

Reply:
@dryajov Pattern with `tryComplete` and polling is introduced because you commented previously that it's a suitable solution
https://github.com/codex-storage/nim-codex/pull/716#discussion_r1526619271

I think however that we _can_ do it with `handle sync`, having a small tradeoff in mind that the main thread can be blocked now and then for a short period of time because chronos signal may arrive earlier than taskpool's flowar item on the internal queue. However it should be negligible. WDYT?

I just need to make sure that the main thread is not going to be a part of the worker threads.